### ISSUE TITLE
 Product changes don't take effect if tax category is "None"

### ISSUE THREAD

**Issue by filipefurtad0** [openfoodfoundation/openfoodnetwork#7170]
## Description

If the instance requires tax categories on products, product edits made under `/admin/products/<product_name>/edit` do not take effect if the selected tax category is 'None'. Attempting to do so shows no warning is shown to the user.

## Expected Behavior

Not sure, I see two options.

If an admin tries to edit a product, in an instance which requires tax categories on products:
- the 'None' option is not selected by default - it is removed from the dropdown list
- if the user selects 'None' an error message is displayed  

## Actual Behaviour

If an admin tries to edit a product, in an instance which requires tax categories on products, with 'None' is selected as tax category the changes don't persist, when clicking 'Update'.  

## Steps to Reproduce

As superadmin:
1. Set an instance to require tax categories on products `/admin/tax_settings/edit`

As admin:
2. Edit a product, visiting /admin/products/<product_name>/edit
3. Choose Tax Category 'None'
4. Make changes to the product.
5. Click 'Update' and see how changes don't take effect; no banned or error message is displayed.
6. Conversely, changes should take effect if any other tax category is chosen.

## Animated Gif/Screenshot

![Peek 2021-03-19 18-20.gif](https://images.zenhubusercontent.com/5d761b049dde650001eaf32f/1b0e5d17-d1f1-4bb1-a4ba-d76864c661d1)

## Workaround

Choosing a tax category other than 'None'

## Severity

s3?

## Your Environment

* Version used: v3.6.0
* Browser name and version: Firefox 86
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04

**Comment by tschumilas** [comment]
Further to this -  it also seems like it is not possible to configure a tax category = none.   So - on OFN-CAN many products currently have tax category = none.   The descriptions of those products cannot be edited presently.   I know this isn't an issue for any other instance,   so keep it an s3 for sure.  I'm going to spend TV time tonight going through and changing all those products to 'zero-rated' so I don't get more user calls.   

It is not possible for a user today  to select a tax category = none.  Trying to do so brings the following error.  So - at least that is good,  because once I change all the 'none'  to 'zero-rated',   there won't be more created.

As an aside -  technically -  there should be both a 'none' and a 'zero-rated' tax category.  Zero Rated - means the goods are taxable,  but the legislator had decided to rate it at '0' (for now).  Someone with zero rated products can claim an input tax credit for the tax they paid to aquire or provide for zero rated goods.  (Almost all food is zero rated here in Canada)  'None'  means the product is exempt from tax altogether.  No input tax credit can be claimed on a 'none' category product.   SO -  our users SHOULD be seleccting 'zero rated'  and not 'none' technically.   (I'm sure they don't know that,  and I'm doing them a favour.) 

Signed,  the tax queen. :)
![d9f9f35b-8a73-451e-a96b-f30501236e00](https://user-images.githubusercontent.com/13171778/111836801-414a6580-88cd-11eb-9e3c-eb73a3915503.png)

**Comment by tschumilas** [comment]
maybe the question here is why does "None" show up in OFN-CAN's dropdown as a tax category selection,  when it is not configured in our tax categories in configuration?     Would there be a way I could make it dissapear as an available tax category?

**Comment by jibees** [comment]
As #7180 is now linked to this issue, please be warn on the fact that if the field `VARIANT UNIT NAME` isn't filled in, the Product doesn't save without any warning. 
Changes required : 
 - Also add feedback for this empty `VARIANT UNIT NAME` field
 - Mark field as required as it:
 
<img width="260" alt="Capture d’écran 2021-04-13 à 11 45 01" src="https://user-images.githubusercontent.com/296452/114532978-b0f20d00-9c4d-11eb-8ffb-8c24f6174777.png">

**Comment by apricot12** [comment]
Okay so I just looked around a bit and there are a few things I've noticed. 
**A)** You cant select tax category = none and save unless you've already selected any one tax category as the default in `http://localhost:3000/admin/tax_categories/#/edit`

**B)** Once you've done the above, you can go back to `/admin/products/<product_name>/edit` and save tax category = none, along with any edits to the product.

![Tax-category](https://user-images.githubusercontent.com/65319144/124106041-7a27d800-da81-11eb-931a-31a30f5dc22a.gif)

**Comment by filipefurtad0** [comment]
Hey @apricot12 ,

That's a very interesting finding, thanks for investigating further :muscle:

So maybe we could simplify and simply remove `https://staging.openfoodnetwork.org.uk/admin/tax_settings/edit` altogether?

We would then keep the "None" tax category available on all dropdowns (as currently is), and enforce that one tax category needs to be default, also to account that there are products which `have tax_category_id` empty. Do we need this?

Maybe @Matt-Yorkley has some ideas on this one as well?

**Comment by apricot12** [comment]
> Hey @apricot12 ,
> 
> That's a very interesting finding, thanks for investigating further 
> 
> So maybe we could remove `https://staging.openfoodnetwork.org.uk/admin/tax_settings/edit` option altogether?
> 
> We would then keep the "None" tax category available on all dropdowns (as currently is), and enforce that one tax category needs to be default.
> 
> Indeed we currently have are products which have `tax_category_id` is empty. Do we need this?
> 
> Maybe @Matt-Yorkley has some ideas on this one as well?

So tax categories will be enabled by default, and then enforce the selection of any one tax category as the default?

**Comment by lin-d-hop** [comment]
I've had a quick look into this. I'm not confident on the below so @filipefurtad0 @RachL would you check if you agree?

My understanding (as background):
- Tax categories are general. They are used to inform which tax rate should be applied at checkout based on buyer and seller locations.
- Even if the producer or person creating the product is not tax registered, someone selling the product could be tax registered. Hence every product should have a tax category.

For this reason I think we should:
1. Remove the super admin option in `/admin/tax_settings/edit`, instead enforcing ticked (all products require tax category) by default.
2. Migrate all products that do not have a tax category set to the 'default' tax category

Currently not all instances have a 'default' tax category set. We need to think about this from a roll-out perspective. We'll need to make this clear in the release notes. For **non-managed instances** the data migration will have to handle the case that a default tax category might not exist. And for **managed instances** we'll need to ensure that all instances have made one of their tax categories 'default' such that there are no unintended consequences. As this is a change that requires a data migration it will be difficult to roll back so we need to make sure we get it right.

Ideally, we'd also want a tooltip to explain why the correct tax category should be set even if the person creating the product is not tax registered.

All of that make this quite a substantial change to solve an S3.

It would be good to understand if resolving this would enable some code removal and simplifications. As this is an S3 I feel that this change might need to be officially prioritised before continuing.

**Comment by RachL** [comment]
I think I agree to the approach but a little voice is my head is also wondering what would happen if we force "none" to be the default tax category everywhere? 

I mean, for instances which do not have default tax category, what's the impact? 

Do we know currently which instance chose what for managed instances?

**Comment by filipefurtad0** [comment]
Agree with the approach :+1: 

> what would happen if we force "none" to be the default tax category everywhere?

Currently "None" means that the `tax_category_id` entry on the DB for that product is empty. So, to remove the superadmin option perhaps we could define "None" as an equivalent of a "zero rated" tax category, and then "force" this to be the default during the data migration. So, in practice, setting all the "None" products as "zero rated".

> Then if no default exists then the migration chooses the first tax category from the list and sets that as default.

I think this could have some undesirable effects, like setting unwanted values as default.

**Comment by RachL** [comment]
> Currently "None" means that the tax_category_id entry on the DB for that product is empty. 

Ah! I didn't know this. I think FR user are treating this as product without added tax :/

**Comment by lin-d-hop** [comment]
> I think this could have some undesirable effects, like setting unwanted values as default.
Obviously this is true, but without this we need to enable the empty tax category in the DB and this also has unwanted side effects - this bug for example.

I think it is fair to make a change to the platform that requires actions to be take first so long as there is sufficient warning. We are doing this with the removal of StripeConnect. I would prefer to take this small step in increase the ability of the platform to enforce correctness. I would prefer this over leaving superfluous options within the platform that increase the risk of future bugs.

**Comment by tschumilas** [comment]
Just explaining why 'zero-rated'  and  'none' are different:     Zero-Rated means the product IS a taxable product -  but the government of the day has decided the tax rate = 0.    In Canada,  because a zero-rated product is a taxable product -  it qualifies for tax credits.  (ie -  I sell broccol which is zero-rated.  But eventhough I don't collect or remit any tax on it,  I get tax credits for all the inputs I used to make the broccoli.)     None is what we have been using to indicate a product is not taxable.  I think in Canada the term is tax except.    It means that no tax is collected or remitted -  but no tax credits apply either.

In practice our users mix these up and have been selecting 'none' when they mean 'zero-rated'.   I'd welcome the opportunity to make 'zero-rated'  a default here -  and have all the 'none' categories change to 'zero-rated'  if we could manage it?   I don't think there are any 'tax except' products on OFN here --- these are really obscure things.

I know it could be different elsewhere -  but just FYI.   

I'm also quite happy to just create a new category called 'tax except' here -   in which case we would never need 'none'.

**Comment by lin-d-hop** [comment]
@tschumilas This is true from a user perspective. But very different from a tech perspective.

None is a blank space in the DB.
Anything else (tax exempt, zer0-rated, tax free etc etc) is a user created tax category actively chosen and selected.

Every product should have a tax rate selected. No product should not a blank space beside it in the DB.
